### DISCUSSION TITLE
Multiple dynamic op outputs

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8485] [Q&A]
Hey folks, new here :wave: I have a question about multiple dynamic op outputs:
If I have a single dynamic output, I would decorate with e.g.
```
@op(out=DynamicOut(str))
def create_inputs():
    for i in range(3):
        yield DynamicOutput(value=f"value{i}", mapping_key=f"output{i}")

@graph
def process(inp: str):
    res = another_op(inp)
    ...

@job
def my_job():
    inputs = create_inputs()
    inputs.map(process)
```
now suppose I want to map to a graph that takes several input arguments, which might get passed to different ops inside

```
@graph
def process(inp1: str, inp2: int, inp3: float):
    res1 = another_op1(inp1)
    res2 = another_op2(inp2)
    res3 = another_op3(inp3)
    ...
```
how would I annotated and implement a `create_inputs` op that would yield "dynamic Tuple[str, int, float]s", s.t. I can still use
```
    inputs = create_inputs()
    inputs.map(process)
```
in my job, together with correct annotations on the graph in dagit (with names and all)? Thanks a bunch :pray:

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1645119661433069?thread_ts=1645119661.433069&cid=C01U954MEER).*

**Comment by yuhan** [comment]
There may not be a way to achieve exactly what you desire.

if you have the dynamic output be `Tuple[str, int, float]` you could just have `process` take the tuple and then add an op that splits it out in to multiple outputs.
```
@graph
def process(tuple_val: Tuple[str, int float]):
    inp1, inp2, inp3 = split(tuple_val)
    res1 = another_op1(inp1)
    res2 = another_op2(inp2)
    res3 = another_op3(inp3)
    ...
```
The `@graph` / `@job` functions dont execute at runtime, only at init time to build the dependency structure, so data manipulation has to happen in an `@op`
